omegaconf/basecontainer.py

```diff
@@ -49,7 +49,7 @@ def _normalize_ref_type(self) -> None:
 
     def _resolve_with_default(
         self,
-        key: Union[str, int, Enum],
+        key: Union[str, int, Enum, float, bool],
```

Comment:
`_resolve_with_default` gets called by `ListConfig.__getitem__`, which makes me think that something like `Union[DictKeyType, ListIndexType]` would be more correct...

Reply:
Should I add an alias `ListIndexType = int` to `base.py`?

Reply:
Alternatively, we could just use Union[DictKeyType, int] without making an alias.
For example, we currently have
```
    @staticmethod
    def _item_eq(
        c1: Container,
        k1: Union[DictKeyType, int],
        c2: Container,
        k2: Union[DictKeyType, int],
    ) -> bool:
```
in `basecontainer.py`.

Reply:
int is in DictKeyType now, right?

This doesn't work?
```
    @staticmethod
    def _item_eq(
        c1: Container,
        k1: DictKeyType,
        c2: Container,
        k2: DictKeyType,
    ) -> bool:
```

Reply:
Yes, `int` is in `DictKeyType`. That does work.
The only point was that some uses cases for `_item_eq` or `_resolve_with_default` will be driven by `ListConfig` (not just by `DictConfig`), so one might worry that using the name "DictKeyType" might be misleading because it mentions only "Dict" and not "List".

Reply:
Maybe it would be appropriate to rename `DictKeyType -> BaseKeyType`, or simply `DictKeyType -> KeyType`?

Reply:
_item_eq and _resolve_with_default are internal functions. I wouldn't worry too much about them (we can even make them take Any if it simplifies things).

> Maybe it would be appropriate to rename DictKeyType -> BaseKeyType, or simply DictKeyType -> KeyType?

Not sure about it, there is a big difference in the key type between DictConfig and ListConfig.
We wouldn't claim that ListConfig can operate on the full spectrum of types that DictConfig can.

I can see why this function makes you uncomfortable. How about we just unroll the union and use explicit types in cases where we need to support both List and Dict key types?

Reply:
Ok, sounds good :)

Reply:
Resolved in https://github.com/omry/omegaconf/pull/484/commits/cac50b21e6394cbacdf94400406b683fe47ec727